docs/module-sources.rst

```diff
@@ -0,0 +1,567 @@
+(equivalent to ``flatpak run --sandbox``) and won't have network, dbus or
+any host access. The ``apply_extra`` can itself be a seperate script file
+too but must be installed as ``${FLATPAK_DEST}/bin/apply_extra``.
+
+``unsquashfs -quiet -no-progress foo.snap`` extracts the snap package
+into a directory called ``squashfs-root``, then the contents of that
+direcory are moved to ``/app/extra`` and lastly ``squashfs-root`` and the
+snap package itself is cleaned up to save space.
+
+This is another example that can be used to extract Debian packages (``.deb``).
+
+.. code-block:: yaml
+
+  - name: foo
+    buildsystem: simple
+    build-commands:
+      - install -Dm755 apply_extra ${FLATPAK_DEST}/bin/apply_extra
+      - install -Dm644 -t ${FLATPAK_DEST}/share/metainfo/ ${FLATPAK_ID}.metainfo.xml
+      - install -Dm644 -t ${FLATPAK_DEST}/share/applications/ ${FLATPAK_ID}.desktop
+      - install -Dm644 -t ${FLATPAK_DEST}/share/icons/hicolor/scalable/apps/ ${FLATPAK_ID}.svg
+      - ln -s ${FLATPAK_DEST}/extra/bin/foo_binary ${FLATPAK_DEST}/bin/foo_binary
+
+    sources:
+      - type: script
+        dest-filename: apply_extra
+        commands:
+          - bsdtar --to-stdout -xf foo.deb data.* | bsdtar -xf -
+          - rm foo.deb
+
+Flatpak extensions can also use extra-data sources in a similar manner.
+This is an example of an extension manifest utilising an extra-data
+source.
+
+.. code-block:: yaml
+
+  app-id: com.foo.app.Plugins.audio
+  runtime: com.foo.app
+  sdk: org.freedesktop.Sdk//23.08
+  build-extension: true
+  separate-locales: false
+  modules:
+    - name: foo
+      buildsystem: simple
+      build-commands:
+        - install -Dm755 apply_extra -t $FLATPAK_DEST/bin/
+      sources:
+        - type: script
+          dest-filename: apply_extra
+          commands:
+            - bsdtar -xf foo.qdz --strip-components=3
+            - rm -f foo.qdz
+        - type: extra-data
+          filename: foo.qdz
+          url: http://example.org/foo_v1.qdz
+          sha256: 2027ed7e5935612a38c3f7b751b6bd23cfb5b06d1a1e5f34907bc021b09de225
+          size: 523360406
+
+For more information on Flatpak extensions please see :doc:`extension`.
+
+Directory source
+----------------
+
+A directory source with ``type: dir`` is best suited for use in CI to
+build the latest changes and avoid downloading multiple times. These
+don't support any caching, so it will be rebuilt each time the
+application is being built.
+
+When submitting an application to software stores like Flathub, ``dir``
+should be avoided altogether.
+
+``path`` should be the path of the local directory relative to the
+manifest root path, whoose contents will be copied during build.
+
+.. code-block:: yaml
+
+  modules:
+    - name: foo
+      buildsystem: simple
+      build-commands:
+        - ...
+      sources:
+        - type: dir
+          path: icons
+
+Additonally there are `bzr`, `svn` and `data` sources supported. Please
```

Comment:
Maybe mention some formats require specific tools are installed. Flathub doesn’t have bzr iirc.

Reply:
I think it's not `data` but `inline`.

Reply:
Fixed